webapi/src/main/scala/org/knora/webapi/responders/v2/SearchResponderV2.scala

```diff
@@ -2556,12 +2556,12 @@ class SearchResponderV2 extends ResponderWithStandoffV2 {
           */
         def getDependentResourceIrisPerMainResource(prequeryResponse: SparqlSelectResponse,
                                                     transformer: NonTriplestoreSpecificConstructToSelectTransformer,
-                                                    mainResourceVar: QueryVariable): Map[IRI, Set[IRI]] = {
+                                                    mainResourceVar: QueryVariable): DependentResourcesPerMainResource = {
 
             // variables representing dependent resources
             val dependentResourceVariablesGroupConcat: Set[QueryVariable] = transformer.getDependentResourceVariablesGroupConcat
 
-            prequeryResponse.results.bindings.foldLeft(Map.empty[IRI, Set[IRI]]) {
+            val dependentResourcesPerMainRes = prequeryResponse.results.bindings.foldLeft(Map.empty[IRI, Set[IRI]]) {
```

Comment:
@benjamingeer  Can I directly create an empty `ErrorHandlingMap` as an accumulator with an apt error message?

Reply:
`foldLeft(new ErrorHandlingMap(Map.empty[IRI, Set[IRI]], { key => throw GravsearchException(s"main resource not found: $key") })) {`?

Reply:
There's no such thing as an empty (or full) `ErrorHandlingMap`, because an `ErrorHandlingMap` is just a wrapper for another map. Why not create the inner map first as the accumulator, and then wrap it in an `ErrorHandlingMap`?

Reply:
I did this and it's fine. But then I moved on and wanted to do the same thing for `Map[IRI, Map[QueryVariable, Set[IRI]]]`. In that case, wouldn't I have to iterate over each entry in the outer map to create an `ErrorHandlingMap` for each `Map[QueryVariable, Set[IRI]]`?

Reply:
You can add the wrapper here:

https://github.com/dhlab-basel/Knora/blob/3cd2123aaa822578e61200740b7a2fe38f60ae31/webapi/src/main/scala/org/knora/webapi/responders/v2/SearchResponderV2.scala#L2648

Like this:

```scala
val valueObjVarToIrisErrorHandlingMap = new ErrorHandlingMap(valueObjVarToIris, { key => throw GravsearchException(s"whatever") })
acc + (mainResIri -> valueObjVarToIrisErrorHandlingMap)
```

Reply:
I think your original idea (use the `ErrorHandlingMap` as the accumulator) might work, too, but I haven't tried it.

Reply:
I tried 
https://github.com/dhlab-basel/Knora/pull/1028#discussion_r234900897 but then it got inconsistent with the accumulator which was still a map.

I had to explicitly annotate `key`. I wonder why.

Reply:
I've experienced that, too, and I don't know why, either. Scala's type inference has some limitations, and for some reason this is one of them.

Reply:
Ok, now I understand. It got inconsistent because I annotated it as `ErrorHandlingMap`. But `ErrorHandlingMap` extends `Map`, so `Map` is fine.